ledger/src/blockstore.rs

```diff
@@ -2712,19 +2712,40 @@ impl Blockstore {
     }
 
     pub fn get_recent_perf_samples(&self, num: usize) -> Result<Vec<(Slot, PerfSample)>> {
-        Ok(self
+        // When reading `PerfSamples` the database may contains samples of both v1 and v2.  We use
+        // size of the binary representation to decide what kind of value it is.  This is fragile,
+        // and will only work as long as the `bincode` serialized representation of type has fixed
+        // size.
+        let samples = self
             .db
             .iter::<cf::PerfSamples>(IteratorMode::End)?
-            .take(num)
             .map(|(slot, data)| {
-                let perf_sample = deserialize(&data).unwrap();
-                (slot, perf_sample)
+                if data.len() == 18 {
+                    let sample: PerfSampleV1 = deserialize(&data)
+                        .expect("PerfSamples column with 18 bytes decodes as `PerfSampleV1`");
+                    (slot, sample.into())
+                } else if data.len() == 26 {
+                    let sample: PerfSampleV2 = deserialize(&data)
+                        .expect("PerfSamples column with 26 bytes decodes as `PerfSampleV1`");
+                    (slot, sample.into())
+                } else {
+                    panic!(
+                        "PerfSamples column should contain either 16 or 26 bytes.  \
+                             Got: {}",
+                        data.len(),
+                    );
+                }
```

Comment:
I understand your desire to use strict length checking, but I do not think it is worth the brittleness of the magic numbers. It will be sufficient to let bincode determine the type (bincode also has additional options we could use, like reject_trailing_bytes, if we want to check for V1 first, for example). Frankly, if the length of bincode serialized representations changes, this column will be the least of our worries.

```suggestion
                if let Ok(sample) = deserialize::<PerfSampleV2>(&data) {
                    (slot, sample.into())
                } else {
                    let sample = deserialize::<PerfSampleV1>(&data)?; // might need to add `.map_err(Into::into)`
                    (slot, sample.into())
                }
```

Reply:
> I understand your desire to use strict length checking, but I do not think it is worth the brittleness of the magic numbers. It will be sufficient to let bincode determine the type (bincode also has additional options we could use, like reject_trailing_bytes, if we want to check for V1 first, for example). Frankly, if the length of bincode serialized representations changes, this column will be the least of our worries.

Not sure, I understand your reasoning here.
`bincode` does not encode any information besides the actual values.
So the only check it can perform is the "EOF" check.
In my mind, checking the size is relatively stable, as, as you are saying yourself, `bincode` can not really change the encoding without breaking everyone.

I can add a unittest that will fail with a meaningful error, if `PerfSampleV1` size changes from 16.
Which should only happen if someone adds or removes fields in the `PerfSampleV1`.
Same for `PerfSampleV2`.

Just reading a value as `PerfSampleV2` and assuming that if it fails, it must be because we really have a `PerfSampleV1` value, seems equally fragile.  As it similarly relies on the size.  The only difference, in my mind, is that this fragility is implicit.  Which, I think, is worth.

But if you insist, I'll change the deserialization to "try V2 and then fallback to V1".

Reply:
> > Just reading a value as PerfSampleV2 and assuming that if it fails, it must be because we really have a PerfSampleV1 value, seems equally fragile. As it similarly relies on the size. The only difference, in my mind, is that this fragility is implicit. Which, I think, is worse.
> 
> Not necessarily.

I do not understand what do you mean by `Not necessarily` here.

>  If we use the `reject_trailing_bytes` that Tyera mentioned (1), we'll get exact length matching since...
> 
>     * Too few bytes will cause deserialize to fail in all cases
> 
>     * Too many bytes with `reject_trailing_bytes` will fail as the property's name would suggest
> 
> 
> Given this, I kind of like Tyera's suggestion of letting bincode do the checking (via the try V2 / fallback to V1) as no magic numbers / additional constants needed.

I do not fully follow your conclusion.
Both approaches work if we only consider v1 and v2.
What I was thinking about is what happens when we change this code in the future.
I would imagine, it is likely that more stats will be added.
In that case, we would need to be able to distinguish more cases.
If we only add stats, we can just do a series of checks.
But if we want to make any other kind of change, reliance on implicit sizes will make it impossible.
There is no way to remove a field, or change a field type.

Explicit sizes do not solve the problem completely, as if a new size happens to match the size of one of the previous versions, the change is still impossible.

At the same time, both approaches break forward compatibility.
If we enable `reject_trailing_bytes` or use explicit sizes – either approach does not allow downgrades, without a restart.

I guess, this would be one of the points for our discussion tomorrow :)

> (1) Incase you weren't aware, the default behavior of bincode Options struct and bincode (de)serialize functions is different in regards to allowing extra bytes, fixed vs. varint, etc. See [this table](https://docs.rs/bincode/1.3.3/bincode/config/index.html#options-struct-vs-bincode-functions).

Yes, it uses fixed sizes for integers.
If we decide to enable `varint`, it would be much harder, if not practically impossible, to use explicit sizes.
Similarly, if a struct contains an `Option` or another kind of enum with values, the length of the bincode encoding starts to depend on the value.